Batched linear system of equations solver (torch.bgesv)

zou3519 opened the pull request:
Implements torch.bgesv, a batched linear system of equations solver.

Adds bindings for MAGMA's gesv_batched function for CUDA.
For CPU, runs THLapack(gesv) in a for loop.

I decided to not build this into `torch.gesv` but if we want to I can change that.

cc @apaszke 

### Test Plan
New unit tests:
- CPU test against `torch.gesv` and specifying outputs
- GPU test for the same thing
- autograd test for derivative

zou3519 committed:
Added bindings for magma gesv_batched

zou3519 committed:
Use for loop for CPU batched gesv

zou3519 committed:
Expose torch.bgesv

zou3519 committed:
Derivative of bgesv

zou3519 committed:
Tests and docs for torch.bgesv

apaszke commented:
cc @fritzo

fritzo commented:
cc @tbrx @dwd31415

apaszke commented on aten/src/TH/generic/THTensorLapack.c:
```diff
@@ -101,6 +101,34 @@ static THTensor *THTensor_(cloneColumnMajor)(THTensor *self, THTensor *src)
+  THTensor_(resizeNd)(self, 3, size, stride);
+  THTensor_(copy)(self, src);
+  return self;
+}
```
I think it'd be better to rearrange some statements in this function. I'd move the `size` and `stride` arrays to the top and used that both to check if `self` matches what we expect, and in a call to `resizeNd` (less duplication, fewer chances for errors).

apaszke commented on test/test_torch.py:
```diff
@@ -2092,6 +2092,37 @@ def test_gesv(self):
+        x_exp, LU_exp = torch.gesv(b.squeeze(0), A.squeeze(0))
+        x, LU = torch.bgesv(b, A)
+        self.assertEqual(x, x_exp.unsqueeze(0))
+        self.assertEqual(LU, LU_exp.unsqueeze(0))
```
Can you make a test that compares to `gesv` in a short loop (let's say we have a 4 elem batch)?

apaszke reviewed (commented):
That looks great, thanks for wrapping this up so quickly! Only a few minor comments.

zou3519 commented on aten/src/THC/generic/THCTensorMathMagma.cu:
```diff
@@ -98,6 +121,76 @@ THC_API void THCTensor_(gesv)(THCState *state, THCTensor *rb_, THCTensor *ra_, T
+  }
+
+  magma_queue_t magma_queue;
+  magma_queue_create_from_cuda(
```
Not yet. Will measure and report back.

zou3519 replied:
The overhead is ~350000 microseconds the first time it's called (on my machine), and ~6 microseconds every time after that. Is that reasonable?

zou3519 committed:
Addressed comments

- Refactored THCTensor_(newBatchedColumnMajor) and
  THTensor_(cloneBatchedColumnMajor)
- In THCTensor_(bgesv), move error checking to after freeing a lot of
  things
- Added a test to check bgesv against gesv (in a loop) for a batch of size 4

zou3519 committed:
Fix typo

colesbury commented:
A few high-level comments:

The equivalent NumPy function is [np.linalg.solve](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.linalg.solve.html)

1) We should prefer one function that handles batched and non-batched computation.
2) We should treat the left-most dimensions as part of the batch. For example, the 4-d tensor `A x B x M x M` should be treated as batch-size `(A x B)`.
3) It would be good to support broadcasting
4) When adding new functions, consider adding them as native functions to ATen.

You don't have to block the PR on these suggestions, but if it's not too much extra work see if you can merge gesv and bgesv so that we don't unnecessarily expand the public API.

zou3519 commented:
Closing this in favor of #4612
